test/integration/auth/auth.prose.test.ts

```diff
@@ -0,0 +1,373 @@
+  });
+
+  describe('SCRAM-SHA-256 and mechanism negotiation', () => {
+    describe('Steps 1-3', function () {
+      const userMap = {
+        sha1: {
+          description: 'user with SCRAM-SHA-1',
+          username: 'sha1',
+          password: 'sha1',
+          mechanisms: ['SCRAM-SHA-1']
+        },
+        sha256: {
+          description: 'user with SCRAM-SHA-256',
+          username: 'sha256',
+          password: 'sha256',
+          mechanisms: ['SCRAM-SHA-256']
+        },
+        both: {
+          description: 'user with SCRAM-SHA-1 and SCRAM-SHA-256',
+          username: 'both',
+          password: 'both',
+          mechanisms: ['SCRAM-SHA-1', 'SCRAM-SHA-256']
+        }
+      };
+      const users = Object.keys(userMap).map(name => userMap[name]);
+      let utilClient: MongoClient;
+
+      /**
+       * Step 1
+       * Create three test users, one with only SHA-1, one with only SHA-256 and one with
+       * both. For example:
+       * db.runCommand(\{createUser: 'sha1', pwd: 'sha1', roles: ['root'], mechanisms: ['SCRAM-SHA-1']\})
+       * db.runCommand(\{createUser: 'sha256', pwd: 'sha256', roles: ['root'], mechanisms: ['SCRAM-SHA-256']\})
+       * db.runCommand(\{createUser: 'both', pwd: 'both', roles: ['root'], mechanisms: ['SCRAM-SHA-1', 'SCRAM-SHA-256']\})
+       */
+      beforeEach(async function () {
+        utilClient = this.configuration.newClient();
+
+        const createUserCommands = users.map(user => ({
+          createUser: user.username,
+          pwd: user.password,
+          roles: ['root'],
+          mechanisms: user.mechanisms
+        }));
+
+        await Promise.all(createUserCommands.map(cmd => utilClient.db('admin').command(cmd)));
+      });
+
+      afterEach(async function () {
+        await Promise.all(users.map(user => utilClient.db('admin').removeUser(user.username)));
+        await utilClient?.close();
+      });
+
+      /**
+       * Step 2
+       * For each test user, verify that you can connect and run a command requiring
+       * authentication for the following cases:
+       *   - Explicitly specifying each mechanism the user supports.
+       *   - Specifying no mechanism and relying on mechanism negotiation.
+       */
+      for (const user of users) {
+        for (const mechanism of user.mechanisms) {
+          it(`authenticates ${user.description} when explicitly specifying ${mechanism}`, {
+            metadata: { requires: { mongodb: '>=3.7.3' } },
+            test: async function () {
+              const options = {
+                auth: {
+                  username: user.username,
+                  password: user.password
+                },
+                authMechanism: mechanism,
+                authSource: 'admin'
+              };
+
+              const client = this.configuration.newClient({}, options);
+              const stats = await client.db('test').stats();
+              expect(stats).to.exist;
+              await client.close();
```

Comment:
this will leak a client if any step between line 97-99 fails.  This might not be a concern because if any of those steps fail, the test run also fails, but we can fix this by closing the client in an `after` hook or just using a finally block

applies to all tests that follow this pattern

Reply:
I've updated all tests to use try...finally for this pattern.